docs/adding-a-library.md

```diff
@@ -0,0 +1,312 @@
+need either git or git-lfs to work, and none of your users will need it either.
+
+```python
+>>> from huggingface_hub import snapshot_download
+>>> snapshot_download("lysandre/arxiv-nlp")
+'/home/lysandre/.cache/huggingface/hub/lysandre__arxiv-nlp.894a9adde21d9a3e3843e6d5aeaaf01875c7fade'
+```
+
+The downloaded files are, once again, cached on your system. The code above will only need to fetch the files the 
+first time it is run; and if you update a file on your repository, then that method will automatically fetch the 
+updated repository next time it is run.
+
+### Finalizing the downstream approach
+
+With the help of these three methods, implementing a download mechanism from the hub should be super simple! Please, 
+feel free to open an issue if you're lost as to how apply it to your library - we'll be happy to help.
+
+## Upstream: creating repositories and uploading files to the hub
+
+The `huggingface_hub` library offers a few tools to make it super simple to create a repository on the hub 
+programmatically, and upload files to that repository. 
+
+We'll see two approaches, that work slightly differently.
+- The `HfApi` class
+- The `Repository` class.
+
+### `HfApi`
+
+The `HfApi` class is a low level class that acts as a thin wrapper around HTTP requests. It has no dependency
+other than `requests`, and can handle repository creation and deletion, repository visibility, model search, as well
+as file uploads.
+
+```python
+>>> from huggingface_hub import HfApi
+>>> api = HfApi()
+```
+
+This class contains a few methods we're interested in: `create_repo` and `upload_file`.
+
+While up to now we didn't need any authorization token to download files from public repositories, we will 
+need one to create a repository and upload files to it. You can retrieve your token using the `HfFolder`:
+
+```python
+>>> from huggingface_hub import HfFolder
+>>> folder = HfFolder()
+>>> token = folder.get_token()
+```
+
+#### `create_repo`
+
+The `create_repo` method may be used to create a repository directly on the model hub. Once you have your 
+token in hand
+
+```python
+>>> from huggingface_hub import HfApi
+>>> api = HfApi()
+>>> api.create_repo(token, "test-model")
+'https://huggingface.co/lysandre/test-model'
+```
+
+You can choose to create repository privately, and to upload it to an organization you are part of.
+
+#### `upload_file`
+
+The `upload_file` method may be used to upload files to a repository directly on the model hub. It needs a 
+token, a path to a file, the final path in the repo, as well as the ID of the repo we're pushing to.
+
+```python
+>>> api.upload_file(
+...	token, 
+...	path_or_fileobj="/home/lysandre/dummy-test/README.md", 
+...	path_in_repo="README.md", 
+...	repo_id="lysandre/test-model"
+... )
+'https://huggingface.co/lysandre/test-model/blob/main/README.md'
+```
+
+### `Repository`
+
+The `Repository` class is thought differently to the `HfApi` class: it is a wrapper over usual git & git-lfs methods.
```

Comment:
This sentence sounds a little awkward. Maybe something like:

```suggestion
The `Repository` class, on the other hand, is a wrapper over git & git-lfs methods.
```

Reply:
> The `Repository` class is the main way to programmatically push models or other repos to huggingface.co, and it is simply a wrapper over git & git-lfs methods.